presto-main/src/main/java/io/prestosql/execution/buffer/PagesSerde.java

```diff
@@ -42,51 +45,73 @@
     private final BlockEncodingSerde blockEncodingSerde;
     private final Optional<Compressor> compressor;
     private final Optional<Decompressor> decompressor;
+    private final Optional<SpillCipher> spillCipher;
 
-    public PagesSerde(BlockEncodingSerde blockEncodingSerde, Optional<Compressor> compressor, Optional<Decompressor> decompressor)
+    public PagesSerde(BlockEncodingSerde blockEncodingSerde, Optional<Compressor> compressor, Optional<Decompressor> decompressor, Optional<SpillCipher> spillCipher)
     {
         this.blockEncodingSerde = requireNonNull(blockEncodingSerde, "blockEncodingSerde is null");
+        checkArgument(compressor.isPresent() == decompressor.isPresent(), "compressor and decompressor must both be present or both be absent");
         this.compressor = requireNonNull(compressor, "compressor is null");
         this.decompressor = requireNonNull(decompressor, "decompressor is null");
-        checkArgument(compressor.isPresent() == decompressor.isPresent(), "compressor and decompressor must both be present or both be absent");
+        this.spillCipher = requireNonNull(spillCipher, "spillCipher is null");
+        checkState(!spillCipher.isPresent() || !spillCipher.get().isDestroyed(), "spillCipher is already destroyed");
     }
 
     public SerializedPage serialize(Page page)
     {
-        SliceOutput serializationBuffer = new DynamicSliceOutput(toIntExact((page.getSizeInBytes() + Integer.BYTES))); // block length is an int
+        SliceOutput serializationBuffer = new DynamicSliceOutput(toIntExact(page.getSizeInBytes() + Integer.BYTES)); // block length is an int
         writeRawPage(page, serializationBuffer, blockEncodingSerde);
+        Slice slice = serializationBuffer.slice();
+        int uncompressedSize = serializationBuffer.size();
+        byte markers = PageCodecMarker.none();
 
-        if (!compressor.isPresent()) {
-            return new SerializedPage(serializationBuffer.slice(), PageCodecMarker.none(), page.getPositionCount(), serializationBuffer.size());
+        if (compressor.isPresent()) {
+            ByteBuffer compressionBuffer = ByteBuffer.allocate(compressor.get().maxCompressedLength(uncompressedSize));
```

Comment:
Generally, we should avoid `ByteBuffer`.  Here is the code from the ORC reader that unwraps a Slice into the underlying `byte[]`, and we can be sure this is heap based buffer since we are creating the stream above:

https://github.com/prestosql/presto/blob/master/presto-orc/src/main/java/io/prestosql/orc/stream/CompressedOrcChunkLoader.java#L129-L133

Reply:
Ok, I'll use the code you linked here for compression as well and encryption. I think, however, that forcing callers of the `Slice` API to use constants from `sun.misc.Unsafe` and blindly casting is a pretty bad code smell. I trust you when you say it's safe to assume these slices are always backed by heap arrays, but obviously that may not always hold.

If these patterns of slice access are common (which it appears they are becoming at least semi-common), then maybe Slice needs`hasArray()`, `arrayOffset()` and `array()` methods to keep the caller from doing these casts. I'm not a huge fan of `ByteBuffer` as an API design, but those methods do give you a sane way to accomplish this and handle non-heap buffers if they're encountered.

Reply:
I like the idea of adding these APIs to Slice. @dain what do you think?

Reply:
Ya, it would be nice to add an optimal dependency on aircompressor and have a simple util to do the for you... maybe another helper for crypto

Reply:
@electrum I've opened https://github.com/airlift/slice/pull/118 to expose the byte array unwrapping methods we discussed here.